Aggressively GC on high memory usage in worker

ogrisel opened the pull request:
This is a possible fix for some of the spurious worker restarts reported in the discussion of dask/dask#2470.

I ran the test script of https://github.com/dask/dask/issues/2470#issuecomment-338117445 with additional logs before and after the call to `gc.collect()` and I confirm that many unnecessary restarts can be avoided this way when memory is tight.

ogrisel commented:
I don't think there is an easy way to add a test for this.

pitrou commented on distributed/worker.py:
```diff
@@ -2174,20 +2175,25 @@ def memory_monitor(self):
+        if frac > min(self.memory_pause_fraction, self.memory_spill_fraction):
+            # Force a GC event so as to get more accurate memory usage
+            # measurements in case GC is lagging.
+            gc.collect()
```
How often is `memory_monitor` executed? `gc.collect` can become quite costly if there are many live objects.

ogrisel replied:
`gc.collect` is only executed when we are above the high memory fraction. This should not incur any significant performance impact when the memory usage is normal (lower than 70% of the limit by default I think).

pitrou replied:
But imagine a worker that's permanently above that fraction. Now you're running the GC over and again (even if it doesn't release any memory)...

ogrisel replied:
Here is the kind of log I got when running the linked snippet with the 1GB memory limit on the worker with additional logs before and after GC:

https://gist.github.com/ogrisel/6cb75c91d515601c4a659675506ea9ff

I had the code running for 10min or so. So it would not get triggered too often but saved a spurious restart.

ogrisel replied:
> But imagine a worker that's permanently above that fraction. Now you're running the GC over and again (even if it doesn't release any memory)...

The default `memory_monitor_interval` is 200ms according to the default value of the kwarg. I could add a 5s cool-down period between two consecutive gc calls but that will make the code more complex / hackish.

pitrou replied:
Well, It wouldn't be extraordinary for `gc.collect` alone to take 200 ms. Running it on such a short period doesn't sound like a good idea.

Also, I wonder why `gc.collect` is supposed to fix your memory issue at all. Do you have cyclic references somewhere?

ogrisel replied:
It seems to reduce the variance of the measurement of the memory usage of the worker and avoid triggering a worker restart when it's not necessary. If the worker get restarted, all the intermediate tasks are canceled and restarted from scratch which can be very expensive and can even cause the whole "macro-task" to never complete.

mrocklin replied:
We should at least call `gc.collect()` when first transitioning to a paused state.  This occurs below after the `if not self.paused` statement.  I'm curious if this suffices to avoid your problem.

mrocklin replied:
We may also want to call gc.collect() regularly when memory is use is high, I agree with @pitrou though that we should probably control this a little bit.  @bluenote10 's solution in https://github.com/dask/distributed/pull/1255 might be worth looking into.

ogrisel replied:
Calling `gc.collect()` in the paused state is apparently not enough. nanny has still a high probability of reaching it's 95% threshold without the worker having the time to see the internal monitory 80% paused threshold first.

I have tried to make nanny less trigger happy by adding a patience counter for consecutive violations but then the nanny leak test tend to fail with my computer close to freezing for a couple of seconds: the nanny OOM killer became too slow to detect the memory leak.

mrocklin replied:
We might call `gc.collect()` *if* we haven't called it fairly recently?  Taking the implementation from #1255 :

```python
class ThrottledGC(object):
    """
    Allows to run throttled garbage collection as a countermeasure to e.g.:
    https://github.com/dask/zict/issues/19
    """
    def __init__(self, min_interval_in_sec=1.0):
        self.min_interval_in_sec = min_interval_in_sec
        self.last_collect = time()
  
    def collect(self, force_gc=False):
        new_time = time()
        if force_gc or new_time - self.last_collect > self.min_interval_in_sec:
            gc.collect()
            self.last_collect = new_time
```

ogrisel replied:
I have a tried a less aggressive / smarter strategy with extensive logs here: f3981a53a81af4ddfef97f8ed58a9cd61e3dd76c . It kind of works but I still get a random OOM after a while.

ogrisel replied:
At least that last option gets rid of the following erroneous: warning I get most often when running the original snippet on master: "Memory use is high but worker has no data to store to disk.  Perhaps some other process leaking memory?"

mrocklin replied:
Nanny OOMs aren't entirely avoidable if your tasks can quickly generate a moderate fraction of your available memory.  It will always be possible that they can conspire to generate enough data to send you over very quickly.

There are costs and benefits to both approaches.

ogrisel replied:
I will implement the `ThrottledGC` in this PR. But I am not sure if should better implement the strategy of this PR or the one in f3981a5.

mrocklin replied:
I don't have any thoughts on which strategy if better.  If you think what you have in https://github.com/dask/distributed/commit/f3981a53a81af4ddfef97f8ed58a9cd61e3dd76c is nicer then go with that.

ogrisel commented:
I broken the tests by editing the other warning messages. I will push a fix.

mrocklin commented:
FYI I'm also tracking down some intermittent failures that have been popping up recently.  I haven't seen these before though.

mrocklin commented:
See also https://github.com/dask/distributed/pull/1255 

cc @bluenote10

ogrisel commented:
The failure on travis in `test_pause_executor` that fails to be paused is probably caused by my change that causes some sort of race condition in this test.

mrocklin commented:
I believe that you can safely ignore Python 2 nanny test failures: https://github.com/dask/distributed/issues/1491

mrocklin commented:
By "nanny test failures" I mean any test in `test_nanny.py` or any test that starts with `@gen_cluster(..., Worker=Nanny)`

ogrisel commented:
I have updated this PR to do less aggressive GC in the worker (only when it has a chance to be beneficial) while also implementing a throttling strategy based on the fraction of the time in GC over the time between 2 consecutive calls.

I run a bunch of long-running tasks that process chunks with a low memory-limit and it seems to work as expected. The frequency of actual GC events dependes on the value of the memory-limit but I no longer get the misleading "Perhaps some other process is leaking memory?" warning. I get the "Pausing" / "Resuming" warnings from time to time as expected. I have added debug logs to make it possible to use grep to check that the gc is actually called from time to time (e.g. every couple of seconds in my scripts, while each gc duration is on the order of 100ms).

mrocklin commented on distributed/utils.py:
```diff
@@ -1000,3 +1001,42 @@ def __init__(self, *args, **kwargs):
+    throttled_gc.collect() does nothing when repeated calls are so costly and
+    so frequent that the thread would spend more than max_in_gc_frac doing GC.
+    """
+    def __init__(self, max_in_gc_frac=0.2, warn_if_longer=10, logger=None):
```
Do you have any thoughts on the default value for `max_in_gc_frac=0.2`?  Spending 20% of our time in garbage collection seems unpleasant.

mrocklin replied:
The `warn_if_longer=` parameter could also use some mention in the docstring

ogrisel replied:
>  Spending 20% of our time in garbage collection seems unpleasant.

It's an upper-bound. In most situations memory-limit is high enough that GC will almost never be called at all.

But in cases where the worker is under heavy memory pressure, I find it better to spend that time in the GC than run the risk to be spuriously killed by nanny for no valid reason. Note that it's not a smoothed average, just based on the last 2 calls to collect.

ogrisel replied:
I will try to change the fraction to 5% and run some experiments.

ogrisel replied:
In the meantime, I have pushed an improvement to the docstring as requested. Let me know what you think.

mrocklin replied:
> It's an upper-bound. In most situations memory-limit is high enough that GC will almost never be called at all.

OK, that makes sense to me. 

> In the meantime, I have pushed an improvement to the docstring as requested. Let me know what you think.

If we're spending 10s in gc then that is probably also something that we should know about.  Thoughts on reducing this to 1s?

pitrou commented on distributed/worker.py:
```diff
@@ -2209,9 +2214,17 @@ def memory_monitor(self):
+                memory = proc.memory_info().rss
+                if total > need and memory > target:
+                    # Issue a GC to ensure that the evicted data is actually
+                    # freed from memory and taken into account by the monitor
```
Is it expected for evicted data to have cyclic references?

ogrisel replied:
I don't think it's going to be the case for typical workloads but users are free to ship whatever data structures they like into `delayed` function calls. Why kind of problems do you anticipate?

pitrou replied:
I simply think it's bad hygiene to add `gc.collect` calls "just in case" :-)

ogrisel replied:
For the record, this is the kind of actual case that was observed by dask-distributed users:

https://github.com/dask/zict/issues/19#issuecomment-313669395

pitrou commented on distributed/utils.py:
```diff
@@ -1000,3 +1001,49 @@ def __init__(self, *args, **kwargs):
+            self.last_collect = collect_start
+            self.last_gc_duration = max(time() - collect_start, 1e-6)
+            if self.last_gc_duration > self.warn_if_longer:
+                self.logger.warning("gc.collect() took %0.3fs. This is usually"
```
Hmm, I'm curious why you think this is useful. Do you know a use case where that warning would have helped?

ogrisel replied:
To warn datascientists that are inadvertently processing a list of 100 million small dicts at a time in each task instead of processing more smaller lists of a couple of million each instead. For instance parsing large JSON files in large chunks instead of small chunks. 

I assume that large GC pauses might make the worker non-responsive. But if you think this is useless we can get rid of this warning.

pitrou replied:
> I assume that large GC pauses might make the worker non-responsive.

It will. But what if using many objects is deliberate and required by the workload? Do we want to warn against any possible programming mistake?

mrocklin replied:
I think that users would appreciate getting a warning if GC is taking seconds.  They have responded well to similar feedback that we have provided in the past (like long event loop pauses).

pitrou commented on distributed/utils.py:
```diff
@@ -1000,3 +1001,49 @@ def __init__(self, *args, **kwargs):
+    to control to log a warning level message whenever an actual call to
+    gc.collect() lasts too long.
+    """
+    def __init__(self, max_in_gc_frac=0.2, warn_if_longer=10, logger=None):
```
Nitpicking, but allowing 20% of the time inside the GC sounds like a lot.

ogrisel replied:
Would 5% be a reasonable default?

pitrou replied:
It certainly sounds better...

ogrisel commented:
I have reworked this PR to address the comments. I now only trigger throttled GC when it's promising, that is after data eviction.

I decided to know longer call the throttled GC after a paused as it's better to call it after eviction, otherwise the throttling mechanism might prevent interesting collections.

mrocklin commented:
Merged.  

At some point I might want to look at also calling `self.gc.collect` when we transition into a paused state.  This requires some care to avoid the throttling mechanism.  We might get around this by just calling the normal `gc.collect()` here, which might be ok given that this should occur infrequently.
